presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+                    .build();
+
+            JoinEnumerationResult leftResult = getJoinSource(
+                    ImmutableList.copyOf(leftSources),
+                    requiredJoinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            if (leftResult.cost.hasUnknownComponents()) {
+                return UNKNOWN_COST_RESULT;
+            }
+            if (leftResult.cost.equals(INFINITE_COST)) {
+                return INFINITE_COST_RESULT;
+            }
+            PlanNode left = leftResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
+            JoinEnumerationResult rightResult = getJoinSource(
+                    ImmutableList.copyOf(rightSources),
+                    requiredJoinSymbols.stream()
+                            .filter(rightSymbols::contains)
+                            .collect(toImmutableList()));
+            if (rightResult.cost.hasUnknownComponents()) {
+                return UNKNOWN_COST_RESULT;
+            }
+            if (rightResult.cost.equals(INFINITE_COST)) {
+                return INFINITE_COST_RESULT;
+            }
+            PlanNode right = rightResult.planNode.orElseThrow(() -> new IllegalStateException("no planNode present"));
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> sortedOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                    .filter(outputSymbols::contains)
+                    .collect(toImmutableList());
+
+            // Cross joins can't filter symbols as part of the join
+            // If we're doing a cross join, use all output symbols from the inputs and add a project node
+            // on top
+            List<Symbol> joinOutputSymbols = sortedOutputSymbols;
+            if (joinConditions.isEmpty() && joinFilters.isEmpty()) {
+                joinOutputSymbols = Stream.concat(left.getOutputSymbols().stream(), right.getOutputSymbols().stream())
+                        .collect(toImmutableList());
+            }
+
+            JoinEnumerationResult result = setJoinNodeProperties(new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    joinConditions,
+                    joinOutputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty()));
+
+            if (!joinOutputSymbols.equals(sortedOutputSymbols)) {
+                PlanNode resultNode = new ProjectNode(idAllocator.getNextId(), result.planNode.get(), identity(sortedOutputSymbols));
+                result = new JoinEnumerationResult(costProvider.getCumulativeCost(resultNode), Optional.of(resultNode));
```

Comment:
Based on yesterdays offline discussion with @electrum and @haozhun 

We should make sure we do not accidentally trigger fetch of statistics for all the partitions at once for the queries that don't actually need this.

For example 

```
SELECT *
FROM partitioned_table
LIMIT 1
```

query from the planner point of scans all the existing partitions, when execution optimizes it to actually read `any` partition, and yield a single row

Since right now we have only a single cost based rule, that optimized joins only - the example above is not representative. However this example is:

```
SELECT *
FROM partition_table_1
INNER JOIN partition_table_2 ON (x)
LIMIT 1
```

For this example, with the current implementation we will fetch the statistics for all the partitions from both, `partition_table_1` and `partition_table_2`. That will basically kill (timeout?) the Metastore. Currently when `EXPLAIN SELECT * FROM table LIMIT 1` is ran - the query fails with the `There was an error communicating with the Hive metastore`. That happens because Presto is trying to fetch all the statistics for all the partitions all at once.

We should think how can we disable the cost based optimizations for the queries that don't have to be optimized.

Also we should probably fix the `MetastoreHiveStatisticsProvider` to fetch the statistics in small batches, and maybe introduce some time limit for statistics fetch to avoid being timed out on simple `EXPLAIN` queries. @findepi @kokosing  what do you think?

Reply:
> ```SELECT * FROM partition_table_1 INNER JOIN partition_table_2 ON (x) LIMIT 1```

Can optimizer know this query won't benefit from optimizations? E.g. in fact, there can be no matches and so this query might be actually as expensive as if it didn't have the LIMIT.

If there are many partitions to query, we could run some kind of sampling:
- gather as much information as possible within certain time frame (configurable, sane default, like 15s)
- extrapolate the information at hand to the partitions we didn't ask about yet (just as we do when there are no stats for certain partitions)

Reply:
That a good idea actually. We can keep fetching until the deviation criteria is not met, or the timeout is not reached.

How about the next algorithm:

- Fetch a batch of statistics for `N` randomly chosen partitions
- Build a `TableStatistics` object based on that batch
- Select next batch of N, create `TableStatistics`, compare with the `TableStatistics` built based on the previous batches. If the deviation of the values is not more than `M` - return approximated statistics. If the deviation is bigger than `M` - keep trying until the timeout is reached. If the timeout is reached - return `UNKNOWN`

In most of the cases it seems to work pretty well, since partitions are often very similar by the content (partitioned by day). 

For the multi column partitioning we may try different sample schemes as an improvement. If the naive sampling solution wouldn't be good enough.

Reply:
I would only add that, besides randomly choosing partitions, we should get fetch first & last.
Reasoning: for time partitioned data, there is natural ordering on partitions, and poking into first&last can reveal the min/max values for correlated non-partition columns.

btw, @arhimondr you raised this concern for this PR, however, we're going into direction where we don't change any of this PR's code and instead do some fix in the Hive connector.
Since CBO is off by default (and the PR doesn't change this), we should be able to proceed with this PR, right? 
I can work on the metastore sampling.

Reply:
https://github.com/prestodb/presto/issues/10853